File: java/src/main/java/vsdk/toolkit/raycasting/stochasticRaytracing/ElementHierarchyState.java
package vsdk.toolkit.raycasting.stochasticRaytracing;

import java.util.ArrayList;

import vsdk.toolkit.common.Error;
import vsdk.toolkit.common.RenderOptions;
import vsdk.toolkit.common.linealAlgebra.Vector3D;
import vsdk.toolkit.skin.Vertex;

@FunctionalInterface
interface REFINE_ACTION {
    Link apply(
        Link link,
        StochasticRadiosityElement rcvtop,
        double[] ur,
        double[] vr,
        StochasticRadiosityElement srctop,
        double[] us,
        double[] vs,
        RenderOptions renderOptions);
}

@FunctionalInterface
interface ORACLE {
    REFINE_ACTION apply(Link link);
}

/**
Global parameters controlling hierarchical refinement
*/
public class ElementHierarchyState {
    public float epsilon; // Determines meshing accuracy
    public int do_h_meshing; // If doing hierarchical meshing
    public float minimumArea; // Minimum allowed element area
    public long nr_elements; // Number of elements
    public long nr_clusters; // Number of clusters
    public int tvertex_elimination; // If doing T-vertex elimination for rendering
    public HierarchyClusteringMode clustering; // Clustering mode, 0 => no clustering
    public ORACLE oracle; // Refinement oracle to be used
    public StochasticRadiosityElement topCluster; // Top cluster element of element hierarchy
    public ArrayList<Vector3D> coords;
    public ArrayList<Vector3D> normals; // Created during element subdivision
    public ArrayList<Vector3D> texCoords; // Created during element subdivision
    public ArrayList<Vertex> vertices;

    public ElementHierarchyState() {
        epsilon = 0.0f;
        do_h_meshing = 0;
        minimumArea = 0.0f;
        nr_elements = 0;
        nr_clusters = 0;
        tvertex_elimination = 0;
        clustering = null;
        oracle = null;
        topCluster = null;
        coords = null;
        normals = null;
        texCoords = null;
        vertices = null;
    }

    public static void setActiveState(ElementHierarchyState state) {
        activeStatePtr = state;
    }

    public static ElementHierarchyState activeState() {
        if ( activeStatePtr == null ) {
            Error.fatal(-1, "ElementHierarchyState::activeState", "Element hierarchy state was not initialized");
        }
        return activeStatePtr;
    }

    private static ElementHierarchyState activeStatePtr = null;
}

File: typescript/src/main/vsdk/toolkit/raycasting/stochasticRaytracing/ElementHierarchyState.ts
import { ArrayList } from "../../../../java/util/ArrayList";
import { Logger as VsdkLogger } from "../../common/logging/Logger";
import { RenderOptions } from "../../common/RenderOptions";
import { Vector3D } from "../../common/linealAlgebra/Vector3D";
import { Vertex } from "../../skin/Vertex";
import { HierarchyClusteringMode } from "./HierarchyClusteringMode";
import { Link } from "./Link";

export type REFINE_ACTION = (
  link: Link,
  rcvtop: any,
  ur: number[],
  vr: number[],
  srctop: any,
  us: number[],
  vs: number[],
  renderOptions: RenderOptions
) => Link;

export type ORACLE = (link: Link) => REFINE_ACTION;

/**
Global parameters controlling hierarchical refinement
*/
export class ElementHierarchyState {
  public epsilon: number;
  public do_h_meshing: number;
  public minimumArea: number;
  public nr_elements: number;
  public nr_clusters: number;
  public tvertex_elimination: number;
  public clustering: HierarchyClusteringMode | null;
  public oracle: ORACLE | null;
  public topCluster: any;
  public coords: ArrayList<Vector3D> | null;
  public normals: ArrayList<Vector3D> | null;
  public texCoords: ArrayList<Vector3D> | null;
  public vertices: ArrayList<Vertex> | null;

  public constructor() {
    this.epsilon = 0.0;
    this.do_h_meshing = 0;
    this.minimumArea = 0.0;
    this.nr_elements = 0;
    this.nr_clusters = 0;
    this.tvertex_elimination = 0;
    this.clustering = null;
    this.oracle = null;
    this.topCluster = null;
    this.coords = null;
    this.normals = null;
    this.texCoords = null;
    this.vertices = null;
  }

  public static setActiveState(state: ElementHierarchyState): void {
    ElementHierarchyState.activeStatePtr = state;
  }

  public static activeState(): ElementHierarchyState {
    if (ElementHierarchyState.activeStatePtr === null) {
      VsdkLogger.fatal(-1, "ElementHierarchyState::activeState", "Element hierarchy state was not initialized");
    }
    return ElementHierarchyState.activeStatePtr!;
  }

  private static activeStatePtr: ElementHierarchyState | null = null;
}
